presentation/components/src/presentation-components/tree/controlled/TreeHooks.ts

```diff
@@ -55,284 +55,285 @@ export interface PresentationTreeNodeLoaderProps extends PresentationTreeDataPro
  * @public
  */
 export interface PresentationTreeNodeLoaderResult {
+  /** Tree node loader to be used with a tree component */
   nodeLoader: PagedTreeNodeLoader<IPresentationTreeDataProvider>;
+
+  /**
+   * Callback for when rendered tree node item range changes. This property should be passed to
+   * [ControlledTree]($ui-components) when property `enableHierarchyAutoUpdate` is `true`.
+   * @alpha
+   */
   onItemsRendered: (items: RenderedItemsRange) => void;
 }
 
 /**
  * Custom hooks which creates PagedTreeNodeLoader with PresentationTreeDataProvider using
  * supplied imodel and ruleset.
- *
  * @public
  */
 export function usePresentationTreeNodeLoader(
   props: PresentationTreeNodeLoaderProps,
 ): PresentationTreeNodeLoaderResult {
-  interface Info {
-    treeModel: MutableTreeModel | undefined;
-  }
+  const dataProviderProps: PresentationTreeDataProviderProps = React.useMemo(
+    () => ({
+      imodel: props.imodel,
+      ruleset: props.ruleset,
+      pagingSize: props.pagingSize,
+      appendChildrenCountForGroupingNodes: props.appendChildrenCountForGroupingNodes,
+      dataSourceOverrides: props.dataSourceOverrides,
+      ruleDiagnostics: props.ruleDiagnostics,
+      devDiagnostics: props.devDiagnostics,
+    }),
+    [
+      props.appendChildrenCountForGroupingNodes,
+      props.dataSourceOverrides,
+      props.devDiagnostics,
+      props.imodel,
+      props.pagingSize,
+      props.ruleDiagnostics,
+      props.ruleset,
+    ],
+  );
 
-  const [info, setInfo] = useState<Info>({ treeModel: undefined });
-  const dataProvider = useDisposable(useCallback(
-    () => new PresentationTreeDataProvider(props),
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [info, ...Object.values(props)],
-  ));
+  const [{ modelSource, rulesetRegistration, dataProvider }, setTreeNodeLoaderState] = useResettableState<TreeNodeLoaderState>(
+    {
+      modelSource: new TreeModelSource(),
+      rulesetRegistration: new RulesetRegistrationHelper(dataProviderProps.ruleset),

```

Comment:
Why`rulesetRegistration` needs to be created here instead of letting `PresentationTreeDataProvider` to take care of it? Does ruleset registration handling inside `PresentationTreeDataProvider` break update workflows?

Also doesn't this cause same ruleset to be registered each time `usePresentationTreeNodeLoader` is invoked?

Reply:
Argh, got bit by the global state again. Thanks for pointing it out, I'll fix the repeated registration issue.

`PresentationTreeDataProvider` is being recreated each time tree hierarchy changes because it memorizes previous calls to `geNodesCount` and `getNodes`. Ruleset registration is independent from hierarchy changes and thus is stored as a separate resource.

Reply:
Fix repeated ruleset registration issue in 870a3fc82bd56d2ce9b55f93b556c7e18c373126.